### DISCUSSION TITLE
Iterate nested array in REST API query

### DISCUSSION THREAD

**Discussion by arnazmursidp** [Budibase/budibase#9917] [Help] [answered]
Hello budibase team, i'd like to ask how to iterate nested array using REST API, as i looked into the discussions lists, i have found several workarounds, but i still have a few confusions

here's the response example

```
[
  {
    "car_brand": "Toyota",
    "published_at": "2015-08-05T08:40:51.620Z",
    "car_model": [
      {
        "model_name": "Swift",
        "total_availability": 2
      },
      {
        "model_name": "Tesla",
        "total_availability": 4
      }
    ]
  }
]
```

the schema value only returned `[Object, Object]`, not the actual value itself.

![image](https://user-images.githubusercontent.com/19486285/223450712-a50b494e-6fb2-401e-8aa9-3847639e747d.png)

i'd like to know how to iterate the `car_model` array, in several discussions, it implies that i need to use the transformers, but i'm not sure how to implement it in my project, i have tried using nested data providers, repeaters, and using embed, but because the schema value of `car_model` returns invalid response, i only get the index value, not the actual object itself, i'll attach the project example that i'm working on.

[project example](https://drive.google.com/file/d/1TebSwPTpmdxkjZEmfrs9slCtT1xO1c6D/view?usp=sharing)

thank you for your help!

**Comment by andz-bb** [comment]
Hey @arnazmursidp - thanks for the detailed question and for sending your app export!

You're correct that you will need to use a Transformer for this. To get an array of `car_model` objects from inside all of your result objects you can use a [JS array flat map function](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flatMap) like this. This works exactly like a standard map function and then flattens the returned array by one level after.

```js
return data.flatMap(x => x.car_model)
```
![SCR-20230307-n39](https://user-images.githubusercontent.com/110921612/223488276-2fe7945c-a8bf-43e7-96b7-63bdbb7e14aa.png)

You can see the returned data structure in the screenshot. More advanced transformer functions are possible if you need to include other properties from the original data objects. If you need any help with those let me know and I'll be happy to have a look 😊

**Reply by arnazmursidp** [reply]
hello @andz-bb thanks for the response, but with the answer you provide, i will lose the car brand data, and i need the all data displayed in one container using the car data provider, like this example:

![image](https://user-images.githubusercontent.com/19486285/223619321-9fc8023f-4d5b-4953-85d8-f9d13d4551a3.png)

how can i iterate the `car_model` using the `car_brand` data provider? do i need new data provider and repeater to iterate `car_model`?

**Comment by ConorWebb96** [comment, marked as answer]
Hey @arnazmursidp,

Thanks for exporting your app. I had a look at your code within the description of the card component. I've supplied screenshots of two different versions ones is es6 and the other is pretty standard both of them will display the brands contained within.

<img width="981" alt="standard" src="https://user-images.githubusercontent.com/126772285/223658961-c726fbac-ac5b-4b7c-ab33-088f0664ad69.png">

<img width="977" alt="es6" src="https://user-images.githubusercontent.com/126772285/223658954-0997bda9-2a9f-4616-9202-f6f0f62b2ae0.png">

<img width="1056" alt="result" src="https://user-images.githubusercontent.com/126772285/223658966-d1b44d21-e977-4c61-a39c-33891f2230de.png">

Hopefully this is what you wanted.

**Reply by arnazmursidp** [reply]
hello @ConorWebb96 thanks for the response! i have tried using the javascript function and although i can iterate the value in the text component, my goal is to try iterate the `car_model` in table component, is that possible?

**Reply by arnazmursidp** [reply]
oh wait, we have found a solution, we tried to flatten all the array into one level, and filter the `car_model` value based on `car_brand`, and make a provider for each key in `card_model` objects

![image](https://user-images.githubusercontent.com/19486285/223667481-f19793be-77c7-46b7-95ce-4d02e3c9ad09.png)

i think this is not the proper solution, but it is feasible
